server/lib/two-factor-authentication/yubikey-otp.ts

```diff
@@ -0,0 +1,41 @@
+import crypto from 'crypto';
+
+import { ApolloError } from 'apollo-server-express';
+
+import User from '../../models/User';
+
+import { Token } from './lib';
+
+export default {
+  async validateToken(user: User, token: Token): Promise<void> {
+    if (!user.yubikeyDeviceId) {
+      throw new Error('User is not configured with YubiKey OTP 2FA');
+    }
+
+    if (token.code.length !== 44) {
+      throw new ApolloError('Two-factor authentication code is invalid', 'INVALID_2FA_CODE');
+    }
+
+    const tokenDeviceId = token.code.substring(0, 12);
+    if (tokenDeviceId !== user.yubikeyDeviceId) {
+      throw new ApolloError('Two-factor authentication code is invalid', 'INVALID_2FA_CODE');
+    }
+
+    const valid = validateYubikeyOTP(token.code);
+    if (!valid) {
+      throw new ApolloError('Two-factor authentication code is invalid', 'INVALID_2FA_CODE');
+    }
+  },
+};
+
+async function validateYubikeyOTP(otp: string): Promise<boolean> {
+  const response = await fetch(
+    `https://api2.yubico.com/wsapi/2.0/verify?${new URLSearchParams({
+      otp,
+      nonce: crypto.randomBytes(16).toString('base64'),
```

Comment:
What's the purpose of this nonce? Are we suppose to read it back in the response?

Reply:
Its a required field on the API, I am not sure if we need to validate it on the response. From what I could gather its used by the API to differentiate between a replayed OTP (otp only) and a replayed request (same OTP and nonce)